### PR TITLE
Add shutdown event for plugins

### PR DESCRIPTION
**Pull request by ixje**
Some plugins might need to clean up resources before `neo-cli` is shutdown, or perhaps they want to trigger some external event. Currently this is not possible, this PR makes that possible.

### CODE REVIEW

File: neo/Plugins/Plugin.cs [hunk 1]
```diff
@@ -62,6 +62,10 @@ protected virtual void OnPluginsLoaded()
         {
         }
 
+        protected virtual void OnShutdown()
```
Comment (erikzhang) [inline neo/Plugins/Plugin.cs]: Should all plugins implement `IDisposable`?

Comment (ixje) [reply neo/Plugins/Plugin.cs]: l think that could achieve the same.

Comment (ixje) [reply neo/Plugins/Plugin.cs]: Do we want to enforce plugins to implement `Dispose()` or provide a default implementation they optionally override if needed?

1. `public abstract void Dispose();`
2. `public void Dispose() {};`

Comment (erikzhang) [reply neo/Plugins/Plugin.cs]: `virtual` is good.

File: neo/NeoSystem.cs [hunk 2]
```diff
@@ -44,6 +44,10 @@ public void Dispose()
         {
             RpcServer?.Dispose();
             EnsureStoped(LocalNode);
+            foreach (var p in Plugin.Plugins)
+            {
+                p.Dispose();
+            }
```
Comment (erikzhang) [inline neo/NeoSystem.cs]: Some plugins may access `RpcServer`. So maybe we need to dispose the plugins first.

**Comment by erikzhang** [comment]
Can someone do some testing?

**Comment by Tommo-L** [comment]
It works fine in my test case. I create a new SystemLogs plugin.

```c#
public override void Dispose()
{
      Console.WriteLine("SystemLog disposed!");
      base.Dispose();
}
```

image

**Comment by Tommo-L** [comment]
I want to create a unit test `UT_Plugins`, but it cann't passed, as the `NeoSystem` contains some single instance object.

```c#
namespace Neo.UnitTests.Plugins
{
    class MyPlugins : Plugin
    {
        public bool Disposed = false;

        public override void Configure()
        {
        }

        public override void Dispose()
        {
            Disposed = true;
        }
    }

    [TestClass]
    [NotReRunnable]
    public class UI_Plugins
    {
        [TestMethod]
        public void TestDispose()
        {
            NeoSystem testBlockchain = TestBlockchain.InitializeMockNeoSystem();
            MyPlugins myPlugins = new MyPlugins();
            Plugin.Plugins.Add(myPlugins);
            testBlockchain.Dispose();
            myPlugins.Disposed.Should().Be(true);
        }
    }
}
```

**Comment by lock9** [comment]
@Tommo-L I'm having different results and I don't know why:
image

I updated the ApplicationLogs plugin. I tried with and without starting neo-cli using --rpc, and in both cases I can read "configured" but not "disposed" 🤔

**Comment by Tommo-L** [comment]
@lock9  Have you replaced the Neo.dll?  It also works fine in my local according to your changes.
image

**Comment by lock9** [comment]
@Tommo-L I forgot to update the reference in the neo-cli project 😐

**Pull request merged**